### PR TITLE
util: make rand a standalone util decoupled from rdsn runtime

### PR DESCRIPTION
**Pull request by neverchanje**
The random number generation API, `dsn_random64`, is tightly coupled in rdsn runtime for historical issues. As a part of the long-term plan of rdsn (see #141 ) we decoupled the RNG to make further refactoring possible.
This PR also includes some unit tests.

### CODE REVIEW

File: include/dsn/utility/rand.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#pragma once
+
+#include <cstdint>
+#include <cstddef>
+#include <limits>
+
+namespace dsn {
+namespace rand {
+
+// uint64n returns, as an uint64_t, a non-negative pseudo-random number in [min, max].
```
Comment (shengofsun) [inline include/dsn/utility/rand.h]: 注视是uint64n, 代码是uint64in。
不过这命名没看懂，n和in代表啥？

File: src/core/core/rand.cc [hunk 2]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) 2017, Xiaomi, Inc.  All rights reserved.
```
Comment (shengofsun) [inline src/core/core/rand.cc]: 文件名还是统一用cpp吧

File: include/dsn/utility/rand.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+inline uint32_t uint32n(uint32_t n) { return static_cast<uint32_t>(uint64n(n)); }
+
+// uint32 returns a pseudo-random 32-bit value as a uint32_t.
+inline uint32_t uint32() { return uint32in(0, std::numeric_limits<uint32_t>::max()); }
+
+// uint64 returns a pseudo-random 64-bit value as a uint64_t.
+inline uint64_t uint64() { return uint64in(0, std::numeric_limits<uint64_t>::max()); }
+
+// float64 returns, as a double, a pseudo-random number in [0.0,1.0].
+inline double float64n() { return uint64in(0, 1000000000) / 1000000000.0; }
```
Comment (shengofsun) [inline include/dsn/utility/rand.h]: float64n的命名反而不如原来的probability好。因为概率一定是[0, 1]的，而float不一定是

**Comment by neverchanje** [comment]
@shengofsun 命名根据 golang 的 math.rand，https://golang.org/pkg/math/rand/ 我可以改

**Pull request merged**